### PR TITLE
Cross SDK Collaboration - Self Diagnostics Proposal

### PR DESCRIPTION
**Pull request by TimothyMothra**
This is a proposal for a new Self-Diagnostics configuration that would be supported by all first-party Application Insights SDKs (DotNet, Java, Node, Python, and also Javascript and Extensions where applicable).

The ask is for all SDK contributors to provide their feedback here.

### CODE REVIEW

File: CrossSdkCollaboration/Self-Diagnostics.md [hunk 1]
```diff
@@ -0,0 +1,100 @@
...
+
+
+### Configuration String
+
+- The configuration string will consist of a list of key-value pairs separated by semicolon:
+`key1=value1;key2=value2;key3=value3`
+- Reserved characters
+    - `=` equal sign
+    - `;` semi colon
+- The configuration string will be case insensative.
```
Comment (trask) [inline CrossSdkCollaboration/Self-Diagnostics.md]: Some values (e.g. paths on linux) may be case-sensitive.

**Comment by TimothyMothra** [comment]
@reyang @Dmitry-Matveev @trask @markwolff @littleaj @lzchen @Mikhail-msft Please Review. :)

File: CrossSdkCollaboration/Self-Diagnostics.md [hunk 2]
```diff
@@ -0,0 +1,100 @@
...
+
+### Schema
+- `Destination`
+    - Values: **File**, **ETW**, **Console**, **None**
+    - The destination will define where logs are sent. The value will define the sub keywords.
+
+To disable logging
+- Destination will be set to None.
+- The Environment Variable would be set to string.Empty.
+- The Environment Variable would be deleted.
```
Comment (trask) [inline CrossSdkCollaboration/Self-Diagnostics.md]: I'm interested in the default (empty/missing env var) being `Destination=Console;Level=ERROR`, which could then be intentionally suppressed using `Destination=None`

File: CrossSdkCollaboration/Self-Diagnostics.md [hunk 3]
```diff
@@ -0,0 +1,100 @@
...
+    
+#### "Destination=File" schema
+All SDKs would support File logging to provide a basic and consistent supportability experience.
+
+- [Optional] `Path`
+    - Value: The full path to a directory that the SDK will have write access.
+    - Default Value: "%TEMP%"
+- [Optional] `Level`
+    - Value: Verbose, Information, Warning, Error
```
Comment (trask) [inline CrossSdkCollaboration/Self-Diagnostics.md]: Debug, Info, Warn, Error 😄

File: CrossSdkCollaboration/Self-Diagnostics.md [hunk 4]
```diff
@@ -0,0 +1,100 @@
...
+    - Default Value: Verbose
```
Comment (trask) [inline CrossSdkCollaboration/Self-Diagnostics.md]: I like default value `Error` 😄

File: CrossSdkCollaboration/Self-Diagnostics.md [hunk 5]
```diff
@@ -0,0 +1,100 @@
...
+    - Reference: https://docs.microsoft.com/en-us/dotnet/api/system.diagnostics.tracing.eventlevel?view=netframework-4.8
+- [Optional] `MaxSize`
+    - Value: Integer specifying max size in either kilobytes or megabytes? **TODO: Needs a decision**
+    - Default: no max size? **TODO: Needs a decision**
```
Comment (trask) [inline CrossSdkCollaboration/Self-Diagnostics.md]: What happens at max size? Does it logging stop? Does it roll over itself? Does it roll over to a new file? How many historical files are kept?

File: CrossSdkCollaboration/Self-Diagnostics.md [hunk 6]
```diff
@@ -0,0 +1,100 @@
...
+
+##### File Name
+`ApplicationInsightsLog_{DateTime.UtcNow.ToInvariantString("yyyyMMdd_HHmmss")}_{process.ProcessName}_{process.Id}.txt`
```
Comment (trask) [inline CrossSdkCollaboration/Self-Diagnostics.md]: Are `{process.ProcessName}` and `{process.Id}` important?

How about `.log` extension?

Comment (TimothyMothra) [reply CrossSdkCollaboration/Self-Diagnostics.md]: For asp.net apps hosted in IIS, the process name will always be "w3wc" so we still need the process id to differentiate. 

I have no opinion regarding the extension. :)

File: CrossSdkCollaboration/Self-Diagnostics.md (hunk 5 above)
Comment (TimothyMothra) [reply CrossSdkCollaboration/Self-Diagnostics.md]: If we're using the datetime stamp in the file name, i think we could just start a new file at a new datetime stamp.

Comment (trask) [reply CrossSdkCollaboration/Self-Diagnostics.md]: Oh yes, I missed that! How many historical files would we keep?

File: CrossSdkCollaboration/Self-Diagnostics.md (hunk 2 above)
Comment (TimothyMothra) [reply CrossSdkCollaboration/Self-Diagnostics.md]: Let's have a conversation about this, which I think is related to the default Level as well.

If we expect that the SDKs regular logging behavior should be determined by the normal configuration process (config file or coded initialization),
Then this Environment Variable based configuration would be an interrupt/override to that normal process. 
Using DotNet as an example, DotNet logs to ETW by default. But sometimes in production, the customer cannot collect those logs so they would provide this override to send logs to a text file instead.

In this same scenario, when we get a support request, my default go-to would be to tell the customer to collect ALL the logs (Level = Verbose). So that scenario is made simpler if the default logging level is Verbose. 

This is just one moth's opinion, and I'm eager to hear your feedback.

Comment (trask) [reply CrossSdkCollaboration/Self-Diagnostics.md]: ok, so empty/missing env var would not disable logging, it would just revert to the SDK configured logging?

File: CrossSdkCollaboration/Self-Diagnostics.md (hunk 5 above)
Comment (TimothyMothra) [reply CrossSdkCollaboration/Self-Diagnostics.md]: That I don't know. I think it would be inappropriate for the SDK to delete log files so my gut says to leave it up to the end-user. 
If the purpose of this is for temporary troubleshooting, then i would not expect there to be very many files.

File: CrossSdkCollaboration/Self-Diagnostics.md (hunk 2 above)
Comment (TimothyMothra) [reply CrossSdkCollaboration/Self-Diagnostics.md]: That's how I envision using this, yes :)

Comment (TimothyMothra) [reply CrossSdkCollaboration/Self-Diagnostics.md]: I can change the wording in the spec

Comment (trask) [reply CrossSdkCollaboration/Self-Diagnostics.md]: Oops, I skimmed too quickly over the `Introduction` where you explain this, makes perfect sense now, sorry! :sweat_smile:

File: CrossSdkCollaboration/Self-Diagnostics.md (hunk 1 above)
Comment (TimothyMothra) [reply CrossSdkCollaboration/Self-Diagnostics.md]: Thanks, that's a good point.
When we worked on connection strings, I liked having all the keywords case insensitive.

Do you think we should
- keep keywords insensitive, and values case sensitive
- or allow only certain values to be case sensitive, such as FilePath?

File: CrossSdkCollaboration/Self-Diagnostics.md (hunk 2 above)
Comment (TimothyMothra) [reply CrossSdkCollaboration/Self-Diagnostics.md]: No worries, I think several people had a similar misunderstanding. 
There's no requirement to do it this way, but this way would be convenient for dotnet so that's my bias :)

File: CrossSdkCollaboration/Self-Diagnostics.md (hunk 3 above)
Comment (TimothyMothra) [reply CrossSdkCollaboration/Self-Diagnostics.md]: Shorter names are fine with me.
For a starting point, I used the levels used by ETW.
Also assuming that there would be an end-goal for more SDKs to support ETW in the future.
ETW ranks them as following: Verbose > Information > Warning > Error > Critical

What do you think?
What does "Debug" mean to you in this context?

Comment (trask) [reply CrossSdkCollaboration/Self-Diagnostics.md]: These are just Java (log4j) names 😄.

Debug == Verbose

I'm good with Verbose, Information, Warning, Error 👍

File: CrossSdkCollaboration/Self-Diagnostics.md (hunk 4 above)
Comment (trask) [reply CrossSdkCollaboration/Self-Diagnostics.md]: Verbose makes more sense now that I read the Introduction :grin:

File: CrossSdkCollaboration/Self-Diagnostics.md (hunk 1 above)
Comment (trask) [reply CrossSdkCollaboration/Self-Diagnostics.md]: This makes sense to me: "allow only certain values to be case sensitive, such as FilePath"

File: CrossSdkCollaboration/Self-Diagnostics.md [hunk 7]
```diff
@@ -0,0 +1,103 @@
...
+### Configuration String
+
+- The configuration string will consist of a list of key-value pairs separated by semicolon:
+`key1=value1;key2=value2;key3=value3`
+- Reserved characters
+    - `=` equal sign
+    - `;` semi colon
+- The configuration string will be case insensitive. 
+- Keys will not be order dependent and are expected to appear only once.
+- The full length will not exceed XXX characters. **TODO: Needs a decision**
```
Comment (reyang) [inline CrossSdkCollaboration/Self-Diagnostics.md]: Need to define the behavior when the configuration string is malformed. This also needs to take compatibility into consideration (e.g. if we added feature / options in the future).

File: CrossSdkCollaboration/Self-Diagnostics.md [hunk 8]
```diff
@@ -0,0 +1,103 @@
...
+### Schema
+
+
+- `Destination`
+    - Values: **File**, **ETW**, **Console**, **None**
+    - Specifying a destination will "turn-on" temporary logging. This can override anything in the SDK's normal configuration. The destination will define where logs are sent and the value will define the sub keywords.
+
+To revert to the SDK's previously configured logging:
+- Destination will be set to None.
+- The Environment Variable would be set to string.Empty.
```
Comment (reyang) [inline CrossSdkCollaboration/Self-Diagnostics.md]: Curious, would environment variable be an empty string?

File: CrossSdkCollaboration/Self-Diagnostics.md (hunk 7 above)
Comment (TimothyMothra) [inline CrossSdkCollaboration/Self-Diagnostics.md]: Regarding the length, I need the spec to be more mature before I can count the expected number of characters. Lets please save this decision for later.

File: CrossSdkCollaboration/Self-Diagnostics.md (hunk 5 above)
Comment (reyang) [reply CrossSdkCollaboration/Self-Diagnostics.md]: Does this `max size` concept apply to a single running instance (e.g. a process), or globally (all processes)?

File: CrossSdkCollaboration/Self-Diagnostics.md [hunk 9]
```diff
@@ -0,0 +1,103 @@
...
+**TO BE DEFINED**
+
+The DotNet SDKs have ETW logging always on. Any changes to this would be a breaking change.
+Java is specifically interested in supporting ETW.
+
+#### "Destination=Console" schema
+
+**TO BE DEFINED**
+
+One option would be to support [Sysinternals DebugView](https://docs.microsoft.com/en-us/sysinternals/downloads/debugview).
```
Comment (reyang) [inline CrossSdkCollaboration/Self-Diagnostics.md]: DbgView is a no go for sure. It is using a global lock and shared memory which will create several perf/security/compliance issue.

File: CrossSdkCollaboration/Self-Diagnostics.md [hunk 10]
```diff
@@ -0,0 +1,103 @@
...
+To revert to the SDK's previously configured logging:
+- Destination will be set to None.
+- The Environment Variable would be set to string.Empty.
+- The Environment Variable would be deleted.
+    
+#### "Destination=File" schema
+All SDKs would support File logging to provide a basic and consistent supportability experience.
+
+- [Optional] `Path`
+    - Value: The full path to a directory that the SDK will have write access.
```
Comment (reyang) [inline CrossSdkCollaboration/Self-Diagnostics.md]: Need to describe the behavior:
1. If the Path provided is illegal.
2. If the Path provided is legal, but not a folder (e.g. a file, a symlink to a file).
3. If the Path provided is legal, but the folder does not exist - are we creating the folder or fail?
4. If we don't have sufficient access to the folder.
5. If we have access, but error happened while writing to the folder (e.g. disk full).

File: CrossSdkCollaboration/Self-Diagnostics.md (hunk 9 above)
Comment (TimothyMothra) [reply CrossSdkCollaboration/Self-Diagnostics.md]: The Java team proposed this option, so i'm going to defer to their input to drive this conversation.

My opinion is that we should to identify client tools that would work with all the SDKs. DebugView is something I've used in the past so that was my suggestion.

File: CrossSdkCollaboration/Self-Diagnostics.md [hunk 11]
```diff
@@ -0,0 +1,103 @@
...
+- [Optional] `Level`
+    - Value: Verbose, Information, Warning, Error
+    - Default Value: Verbose
+    - Reference: https://docs.microsoft.com/en-us/dotnet/api/system.diagnostics.tracing.eventlevel?view=netframework-4.8
+- [Optional] `MaxSize`
+    - Value: Integer specifying max size in either kilobytes or megabytes? **TODO: Needs a decision**
+    - Default: no max size? **TODO: Needs a decision**
+
+##### File Name
+`ApplicationInsightsLog_{DateTime.UtcNow.ToInvariantString("yyyyMMdd_HHmmss")}_{process.ProcessName}_{process.Id}.log`
```
Comment (reyang) [inline CrossSdkCollaboration/Self-Diagnostics.md]: This is nice!
It'll be great to describe what's the semantic of datetime (e.g. I would be surprised if it is the *SDK initialization time*).

Where do we want to add the metadata? (e.g. the log level and max size):
1. It could be part of the file name (though personally I'd prefer not).
2. It could be the very first line of the log file, e.g. `Internal diagnostic enabled with APPLICATIONINSIGHTS_SELF_DIAGNOSTICS=foo;bar`.

File: CrossSdkCollaboration/Self-Diagnostics.md (hunk 8 above)
Comment (TimothyMothra) [reply CrossSdkCollaboration/Self-Diagnostics.md]: It looks like empty.string is a supported value, but not all tools will allow a user to set this value. 

I was trying to solve the problem of if a user doesn't have access to delete an environment variable. I think the first option "None" will meet this need so I don't think we need to support this third option. I will remove it.

File: CrossSdkCollaboration/Self-Diagnostics.md [hunk 12]
```diff
@@ -0,0 +1,103 @@
...
+
+This format is proposed because multiple SDKs may read the system's environment variable. This format will help users identify the program or process writing to this file.
```
Comment (reyang) [inline CrossSdkCollaboration/Self-Diagnostics.md]: We might need to rephrase `system's environment variable`. I guess we want to use application environment variable?

File: CrossSdkCollaboration/Self-Diagnostics.md [hunk 13]
```diff
@@ -0,0 +1,103 @@
...
+`Destination=File`
+
+##### Commitments
+- DotNet: Yes
+- Java: Yes
+- JavaScript: Not Applicable
+- Node: ???
+- Python: ???
+- App Service Extensions: ???
+
```
Comment (reyang) [inline CrossSdkCollaboration/Self-Diagnostics.md]: nit: remove extra empty lines

File: CrossSdkCollaboration/Self-Diagnostics.md (hunk 9 above)
Comment (trask) [reply CrossSdkCollaboration/Self-Diagnostics.md]: `Console` == `stdout` (e.g. in Java `System.out`)

Comment (TimothyMothra) [reply CrossSdkCollaboration/Self-Diagnostics.md]: most dotnet applications don't run hosted by a console so i don't know what this would mean for us. we would need to identify a tool that could subscribe to Trace.WriteLine

Comment (trask) [reply CrossSdkCollaboration/Self-Diagnostics.md]: makes sense, this could be optional

Comment (TimothyMothra) [reply CrossSdkCollaboration/Self-Diagnostics.md]: Yeah, i'm only asking SDKs to commit to the file based logging. everything else would be optional, but would be great if our flagship SDKs could support as much as possible!

File: CrossSdkCollaboration/Self-Diagnostics.md (hunk 3 above)
Comment (TimothyMothra) [reply CrossSdkCollaboration/Self-Diagnostics.md]: Should "Debug" and "Verbose" be interchangeable?
I don't want to force all SDKs to do things the dotnet way.
But I want to find the right balance between flexibility and consistency.

File: CrossSdkCollaboration/Self-Diagnostics.md (hunk 5 above)
Comment (TimothyMothra) [reply CrossSdkCollaboration/Self-Diagnostics.md]: every sdk and process that reads this configuration string should treat it equally. 
Is there some way I can rephrase this to make it more clear in the spec?

Comment (TimothyMothra) [reply CrossSdkCollaboration/Self-Diagnostics.md]: I've updated the spec with all of these details.

File: CrossSdkCollaboration/Self-Diagnostics.md (hunk 11 above)
Comment (TimothyMothra) [reply CrossSdkCollaboration/Self-Diagnostics.md]: Good idea! I like option #2. This will help a user determine where these files are coming from and more importantly, how to turn them off.

File: CrossSdkCollaboration/Self-Diagnostics.md (hunk 12 above)
Comment (TimothyMothra) [reply CrossSdkCollaboration/Self-Diagnostics.md]: I believe I misspoke here. We shouldn't dictate which environment variable to use. The libraries won't expose this to us. From the SDK's perspective, we read an environment varible name and the OS resolves which one we recieve (process, user, system).

Comment (reyang) [reply CrossSdkCollaboration/Self-Diagnostics.md]: nit: recieve -> receive

File: CrossSdkCollaboration/Self-Diagnostics.md (hunk 10 above)
Comment (TimothyMothra) [reply CrossSdkCollaboration/Self-Diagnostics.md]: If the path is illegal (for any reason) we're going to fall back to the default value. 

Regarding the symlink problem, I've been told this has been fixed in the latest versions of windows but I haven't personally verified this. In my opinion, this is out of scope for the SDK and needs to be addressed by dotnet.

File: CrossSdkCollaboration/Self-Diagnostics.md [hunk 14]
```diff
@@ -0,0 +1,127 @@
...
+- `Destination`
+    - Values: **File**, **ETW**, **Console**, **None**
+    - Specifying a destination will "turn-on" temporary logging. This can override anything in the SDK's normal configuration. The destination will define where logs are sent and the value will define the sub keywords.
+    - If this value is not found or is invalid, the SDK will take no action.
+
+The SDK will revert back to the previous configuration if any of the following conditions is met:
+- The Environment Variable `APPLICATIONINSIGHTS_SELF_DIAGNOSTICS` is deleted.
+- The Environment Variable is set to `Destination=None`.
+- The Environment Variable is illegal or cann ot be parsed.
```
Comment (markwolff) [inline CrossSdkCollaboration/Self-Diagnostics.md]:
```suggestion
- The Environment Variable is illegal or cannot be parsed.
```

**Comment by TimothyMothra** [comment]
Thank you everyone for your feedback. I'm going to merge this as is. 
There are two remaining TODOs that need some measurement to make a decision. 
I'm going to be working on an implementation in the C# SDK and I'll do some testing to make that decision.
